Mathlib/Analysis/InnerProductSpace/Adjoint.lean

```diff
@@ -308,18 +308,36 @@ theorem conj_starProjection {T : E →L[𝕜] E} (hT : IsSelfAdjoint T)
 
 end IsSelfAdjoint
 
+namespace ContinuousLinearMap
+
+variable {T : E →L[𝕜] E} [CompleteSpace E]
+
 open ContinuousLinearMap in
-theorem IsIdempotentElem.hasOrthogonalProjection_range [CompleteSpace E]
-    {p : E →L[𝕜] E} (hp : IsIdempotentElem p) : (LinearMap.range p).HasOrthogonalProjection :=
-  have := hp.isClosed_range.completeSpace_coe
-  .ofCompleteSpace _
+theorem IsIdempotentElem.adjoint_range_eq_range_of (hT : IsIdempotentElem T)
+    (h : (LinearMap.range T)ᗮ = LinearMap.ker T) :
+    LinearMap.range (adjoint T) = LinearMap.range T := by
+  have := hT.hasOrthogonalProjection_range
+  have := hT.star.hasOrthogonalProjection_range
+  rw [← Submodule.orthogonal_orthogonal (LinearMap.range (adjoint T)),
+    orthogonal_range, adjoint_adjoint, ← h, Submodule.orthogonal_orthogonal]
+
+/-- An idempotent operator `T` is self-adjoint iff `(range T)ᗮ = ker T`. -/
+theorem IsIdempotentElem.isSelfAdjoint_iff_orthogonal_range (h : IsIdempotentElem T) :
+    IsSelfAdjoint T ↔ (LinearMap.range T)ᗮ = LinearMap.ker T :=
+  ⟨fun hT => hT.isSymmetric.orthogonal_range, fun h1 => isSelfAdjoint_iff'.mp
+    (coe_inj.mp (LinearMap.IsIdempotentElem.ext
+    (congr(LinearMapClass.linearMap $(h.star.eq))) (congr(LinearMapClass.linearMap $(h.eq)))
+    (adjoint_range_eq_range_of h h1) (orthogonal_range T ▸ h1)))⟩
+
+end ContinuousLinearMap
```

Comment:
To follow up with the previous idea, you can then provide an extensionality lemma for linear star projections. That is,
```lean
lemma ContinuousLinearMap.IsStarProjection.ext_iff {S T : E →L[𝕜] E}
    (hS : IsStarProjection S) (hT : IsStarProjection T) :
    S = T ↔ range S = range T :=
  sorry

alias ⟨_, ContinuousLinearMap.IsStarProjection.ext⟩ := ContinuousLinearMap.IsStarProjection.ext_iff
```
(and I guess a `LinearMap` version too with a `FiniteDimensional` hypothesis.)